File: frontend/mobile/lib/api/shipping.dart
// ignore_for_file: unused_import
library shipping;

import 'dart:convert';
import 'ftl_client.dart';
import 'currency.dart' as currency;
import 'cart.dart' as cart;


class Address {
  String streetAddress;
  String city;
  String state;
  String country;
  int zipCode;

  Address({  required this.streetAddress,  required this.city,  required this.state,  required this.country,  required this.zipCode,  });

  Map<String, dynamic> toMap() {
    return {
      'streetAddress': ((dynamic v) => v)(streetAddress),
      'city': ((dynamic v) => v)(city),
      'state': ((dynamic v) => v)(state),
      'country': ((dynamic v) => v)(country),
      'zipCode': ((dynamic v) => v)(zipCode),
    };
  }

  factory Address.fromMap(Map<String, dynamic> map) {
    return Address(
      streetAddress: ((dynamic v) => v)(map['streetAddress']),
      city: ((dynamic v) => v)(map['city']),
      state: ((dynamic v) => v)(map['state']),
      country: ((dynamic v) => v)(map['country']),
      zipCode: ((dynamic v) => v)(map['zipCode']),
    );
  }

  String toJson() => json.encode(toMap());

  factory Address.fromJson(String source) => Address.fromMap(json.decode(source));
}

class ShippingRequest {
  Address address;
  List<cart.Item> items;

  ShippingRequest({  required this.address,  required this.items,  });

  Map<String, dynamic> toMap() {
    return {
      'address': ((dynamic v) => v.toMap())(address),
      'items': ((dynamic v) => v.map((v) => cart.Item.fromMap(v)).cast<cart.Item>().toList())(items),
    };
  }

  factory ShippingRequest.fromMap(Map<String, dynamic> map) {
    return ShippingRequest(
      address: ((dynamic v) => Address.fromMap(v))(map['address']),
      items: ((dynamic v) => v.map((v) => cart.Item.fromMap(v)).cast<cart.Item>().toList())(map['items']),
    );
  }

  String toJson() => json.encode(toMap());

  factory ShippingRequest.fromJson(String source) => ShippingRequest.fromMap(json.decode(source));
}

class ShipOrderResponse {
  String id;

  ShipOrderResponse({  required this.id,  });

  Map<String, dynamic> toMap() {
    return {
      'id': ((dynamic v) => v)(id),
    };
  }

  factory ShipOrderResponse.fromMap(Map<String, dynamic> map) {
    return ShipOrderResponse(
      id: ((dynamic v) => v)(map['id']),
    );
  }

  String toJson() => json.encode(toMap());

  factory ShipOrderResponse.fromJson(String source) => ShipOrderResponse.fromMap(json.decode(source));
}


class ShippingClient {
  final FTLHttpClient ftlClient;

  ShippingClient({required this.ftlClient});


  Future<currency.Money> getQuote(ShippingRequest request) async {
    final response = await ftlClient.post('/shipping/quote', request: request.toMap());
    if (response.statusCode == 200) {
      return currency.Money.fromJson(response.body);
    } else {
      throw Exception('Failed to get getQuote response');
    }
  }

  Future<ShipOrderResponse> shipOrder(ShippingRequest request) async {
    final response = await ftlClient.post('/shipping/ship', request: request.toMap());
    if (response.statusCode == 200) {
      return ShipOrderResponse.fromJson(response.body);
    } else {
      throw Exception('Failed to get shipOrder response');
    }
  }
}

File: backend/online-boutique/services/shipping/shipping.go
//ftl:module shipping
package shipping

import (
	"context"
	"fmt"
	"math"

	"ftl/cart"
	"ftl/currency"
)

type Address struct {
	StreetAddress string
	City          string
	State         string
	Country       string
	ZipCode       int
}

type ShippingRequest struct {
	Address Address
	Items   []cart.Item
}

//ftl:verb
//ftl:ingress POST /shipping/quote
func GetQuote(ctx context.Context, req ShippingRequest) (currency.Money, error) {
	return moneyFromUSD(8.99), nil
}

type ShipOrderResponse struct {
	ID string
}

//ftl:verb
//ftl:ingress POST /shipping/ship
func ShipOrder(ctx context.Context, req ShippingRequest) (ShipOrderResponse, error) {
	baseAddress := fmt.Sprintf("%s, %s, %s", req.Address.StreetAddress, req.Address.City, req.Address.State)
	return ShipOrderResponse{ID: createTrackingID(baseAddress)}, nil
}

func moneyFromUSD(value float64) currency.Money {
	units, fraction := math.Modf(value)
	return currency.Money{
		CurrencyCode: "USD",
		Units:        int(units),
		Nanos:        int(math.Trunc(fraction * 10000000)),
	}
}
